presto-main/src/main/java/io/prestosql/sql/planner/sanity/PartitionFilterChecker.java

```diff
@@ -0,0 +1,82 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.sanity;
+
+import com.google.common.base.VerifyException;
+import io.prestosql.Session;
+import io.prestosql.execution.warnings.WarningCollector;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.metadata.TableHandle;
+import io.prestosql.metadata.TableProperties;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.sql.planner.TypeAnalyzer;
+import io.prestosql.sql.planner.TypeProvider;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.PlanVisitor;
+import io.prestosql.sql.planner.plan.TableScanNode;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static io.prestosql.SystemSessionProperties.isQueryPartitionFilterRequired;
+
+public class PartitionFilterChecker
+        implements PlanSanityChecker.Checker
+{
+    @Override
+    public void validate(PlanNode plan, final Session session, final Metadata metadata, TypeAnalyzer typeAnalyzer, TypeProvider types, WarningCollector warningCollector)
+    {
+        if (!isQueryPartitionFilterRequired(session)) {
+            return;
+        }
+        plan.accept(new PlanVisitor<Void, Void>()
+        {
+            @Override
+            protected Void visitPlan(PlanNode node, Void context)
+            {
+                for (PlanNode source : node.getSources()) {
+                    source.accept(this, context);
+                }
+                return null;
+            }
+
+            @Override
+            public Void visitTableScan(TableScanNode node, Void context)
+            {
+                TableHandle table = node.getTable();
+                List<ColumnHandle> partitionColumns = getPartitionColumns(table);
+                if (!partitionColumns.isEmpty() && node.getEnforcedConstraint().isAll()) {
+                    String partitionColumnNames = partitionColumns.stream().map(n -> metadata.getColumnMetadata(session, table, n).getName()).collect(Collectors.joining(","));
+                    throw new VerifyException("Filter on partition column required: " + partitionColumnNames);
+                }
+                return null;
+            }
+
+            private List<ColumnHandle> getPartitionColumns(TableHandle handle)
+            {
+                TableProperties properties = metadata.getTableProperties(session, handle);
+                if (properties.getTablePartitioning().isPresent()
+                        && !properties.getTablePartitioning().get().getPartitioningColumns().isEmpty()) {
+                    return properties.getTablePartitioning().get().getPartitioningColumns();
+                }
+                if (properties.getDiscretePredicates().isPresent()
```

Comment:
There are a couple of issues with this check:
- `getDiscretePredicates` is not a well-supported API. We're trying to limit its scope of use and eventually get rid of it.
- As we continue to evolve the pushdown APIs (`applyXXX`), `getTableProperties` may not be able to produce a list of "discrete predicates" or table partitioning for a given "TableHandle". Historically, a TableHandle represented a base table, but that's no longer the case, as it potentially represents a base table with a number of operations applied to it (i.e., a form of virtual or derived table).